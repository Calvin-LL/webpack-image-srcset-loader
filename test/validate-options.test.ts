import webpack from "webpack";

import compile from "./helpers/compile";
import getCompiler from "./helpers/getCompiler";

describe("validate options", () => {
  const tests = {
    sizes: {
      success: [["10w", "1x", "2x", null]],
      failure: [["10.0w"], ["10h"], ["0.5x"], ["0.5"], ["10"], [3]],
    },
    scaleUp: {
      success: [true, false],
      failure: ["true"],
    },
    customOptionsFactory: {
      success: [() => ({})],
      failure: [true],
    },
    esModule: {
      success: [true, false],
      failure: ["true"],
    },
  };

  function createTestCase(
    key: string,
    value: any,
    type: "success" | "failure"
  ) {
    test(`should ${
      type === "success" ? "successfully validate" : "throw an error on"
    } the "${key}" option with ${JSON.stringify(value)} value`, async () => {
      const compiler = getCompiler({ sizes: [null], [key]: value });

      let stats;

      try {
        stats = await compile(compiler);
      } finally {
        if (type === "success") {
          expect((stats as webpack.Stats).hasErrors()).toBe(false);
        } else if (type === "failure") {
          const errors = (stats as webpack.Stats).compilation.errors;

          expect(errors).toHaveLength(1);
          expect(() => {
            throw new Error(errors[0].error.message);
          }).toThrowErrorMatchingSnapshot();
        }
      }
    }, 60000);
  }

  for (const [key, values] of Object.entries(tests)) {
    for (const type of Object.keys(values) as ("success" | "failure")[]) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
