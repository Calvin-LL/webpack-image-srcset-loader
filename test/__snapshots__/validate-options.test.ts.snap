// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate options should throw an error on the "customOptionsFactory" option with true value 1`] = `
"Invalid options object. Image SrcSet Loader has been initialized using an options object that does not match the API schema.
 - options.customOptionsFactory should be an instance of function.
   -> A function that returns the option to be passed on to the next loader."
`;

exports[`validate options should throw an error on the "esModule" option with "true" value 1`] = `
"Invalid options object. Image SrcSet Loader has been initialized using an options object that does not match the API schema.
 - options.esModule should be a boolean.
   -> By default, webpack-image-placeholder-loader generates JS modules that use the ES modules syntax. Default to true."
`;

exports[`validate options should throw an error on the "scaleUp" option with "true" value 1`] = `
"Invalid options object. Image SrcSet Loader has been initialized using an options object that does not match the API schema.
 - options.scaleUp should be a boolean.
   -> Whether or not to scale up the image when the desired size is larger than the image size. Default to false."
`;

exports[`validate options should throw an error on the "sizes" option with ["0.5"] value 1`] = `
"(Emitted value instead of an instance of Error) Invalid options object. Image SrcSet Loader has been initialised using an options object that does not match the API schema.
      - options.sizes ["0.5"] is not a valid value."
`;

exports[`validate options should throw an error on the "sizes" option with ["0.5x"] value 1`] = `
"(Emitted value instead of an instance of Error) Invalid options object. Image SrcSet Loader has been initialised using an options object that does not match the API schema.
      - options.sizes ["0.5x"] is not a valid value."
`;

exports[`validate options should throw an error on the "sizes" option with ["10"] value 1`] = `
"(Emitted value instead of an instance of Error) Invalid options object. Image SrcSet Loader has been initialised using an options object that does not match the API schema.
      - options.sizes ["10"] is not a valid value."
`;

exports[`validate options should throw an error on the "sizes" option with ["10.0w"] value 1`] = `
"(Emitted value instead of an instance of Error) Invalid options object. Image SrcSet Loader has been initialised using an options object that does not match the API schema.
      - options.sizes ["10.0w"] is not a valid value."
`;

exports[`validate options should throw an error on the "sizes" option with ["10h"] value 1`] = `
"(Emitted value instead of an instance of Error) Invalid options object. Image SrcSet Loader has been initialised using an options object that does not match the API schema.
      - options.sizes ["10h"] is not a valid value."
`;

exports[`validate options should throw an error on the "sizes" option with [3] value 1`] = `
"Invalid options object. Image SrcSet Loader has been initialized using an options object that does not match the API schema.
 - options.sizes[0] should be one of these:
   string | null
   Details:
    * options.sizes[0] should be a string.
    * options.sizes[0] should be a null."
`;
